### PR TITLE
[Validator] fix comparisons with null values at property paths

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #29831
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: src/Symfony/Component/Validator/Constraints/GreaterThanOrEqualValidator.php [hunk 1]
```diff
@@ -24,7 +24,7 @@ class GreaterThanOrEqualValidator extends AbstractComparisonValidator
      */
     protected function compareValues($value1, $value2)
     {
-        return $value1 >= $value2;
+        return null !== $value2 && $value1 >= $value2;
```
Comment (sgehrig) [inline src/Symfony/Component/Validator/Constraints/GreaterThanOrEqualValidator.php]:
```suggestion
        return null === $value2 || $value1 >= $value2;
```

File: src/Symfony/Component/Validator/Constraints/GreaterThanValidator.php [hunk 2]
```diff
@@ -24,7 +24,7 @@ class GreaterThanValidator extends AbstractComparisonValidator
      */
     protected function compareValues($value1, $value2)
     {
-        return $value1 > $value2;
+        return null !== $value2 && $value1 > $value2;
```
Comment (sgehrig) [inline src/Symfony/Component/Validator/Constraints/GreaterThanValidator.php]:
```suggestion
        return null === $value2 || $value1 > $value2;
```

File: src/Symfony/Component/Validator/Constraints/LessThanOrEqualValidator.php [hunk 3]
```diff
@@ -24,7 +24,7 @@ class LessThanOrEqualValidator extends AbstractComparisonValidator
      */
     protected function compareValues($value1, $value2)
     {
-        return $value1 <= $value2;
+        return null !== $value2 && $value1 <= $value2;
```
Comment (sgehrig) [inline src/Symfony/Component/Validator/Constraints/LessThanOrEqualValidator.php]:
```suggestion
        return null === $value2 || $value1 <= $value2;
```

File: src/Symfony/Component/Validator/Constraints/LessThanValidator.php [hunk 4]
```diff
@@ -24,7 +24,7 @@ class LessThanValidator extends AbstractComparisonValidator
      */
     protected function compareValues($value1, $value2)
     {
-        return $value1 < $value2;
+        return null !== $value2 && $value1 < $value2;
```
Comment (sgehrig) [inline src/Symfony/Component/Validator/Constraints/LessThanValidator.php]:
```suggestion
        return null === $value2 || $value1 < $value2;
```

**Comment by fabpot** [comment]
@xabbuh Can you finish this one?

**Comment by xabbuh** [comment]
I think we need to make a decision first how we see `null` values should be handled. IMO comparing `null` with greater or less than operators doesn't make any sense.

**Comment by fabpot** [comment]
@xabbuh I agree that comparing `null` with these operators do not make sense.

**Comment by castroCrea** [comment]
?

**Comment by xabbuh** [comment]
@symfony/mergers How do we proceed here? Does a comparison with `null` always lead to a `true` or a `false` result? Throwing an exception would be logical, but isn't possible without breaking backwards compatibility.

**Comment by ogizanagi** [comment]
> How do we proceed here? Does a comparison with null always lead to a true or a false result? Throwing an exception would be logical, but isn't possible without breaking backwards compatibility.

To me null values should be ignored. If any is encountered, we should skip the comparison as it makes no sense.
Throwing is not required if the property is allowed to be null. If it isn't, the property `@NotNull` constraints should have been set in the first place.

**Comment by xabbuh** [comment]
> Throwing is not required if the property is allowed to be null. If it isn't, the proper `@NotNull` constraint should have been set in the first place.

Well, this is not about the property the constraint is attached too. But if you compare it to another property that one can be `null` too.

But to make it a bit more clear what I mean: My reasoning would be that a `null` value should never raise a validation constraint and the comparison constraints should reflect that.

**Comment by sgehrig** [comment]
Just my two cents because I originally raised the issue (#29831). 

> I'd assume that, if the value returned by `propertyPath` is `null`, the validator will just ignore the comparison. Just as the validator ignores the comparison if the *owning* side value (that's where the constraint is attached to) is null.
> - https://github.com/symfony/symfony/issues/29831#issuecomment-453438798

But this may not be correct when dealing with the `Equal`/`NotEqual`/`IdenticalTo`/`NotIdenticalTo` as pointed out by @xabbuh in https://github.com/symfony/symfony/issues/29831#issuecomment-453443926.

So I think to address this correctly, there needs to be a different handling for the equal/not-equal constraints. But for the less-than/greater-than (-equals) constraints a comparison with `null` should be ignored.

**Comment by xabbuh** [comment]
@sgehrig Yes, that pretty much sums up my thoughts.

**Comment by xabbuh** [comment]
Status: Needs Review

**Comment by xabbuh** [comment]
ping @symfony/mergers

**Review by yceruto — APPROVED** [review]

**Review by nicolas-grekas — APPROVED** [review]

**Review by ogizanagi — APPROVED** [review]

**Pull request merged**